File: app/src/main/java/ch/unstable/ost/api/model/impl/Connection.java
package ch.unstable.ost.api.model.impl;

import android.os.Parcel;
import android.os.Parcelable;

import java.util.Date;

import ch.unstable.ost.utils.ParcelUtils;

import static ch.unstable.ost.utils.ObjectsCompat.requireNonNull;

public class Connection implements Parcelable {

    public static final Creator<Connection> CREATOR = new Creator<Connection>() {
        @Override
        public Connection createFromParcel(Parcel in) {
            return new Connection(in);
        }

        @Override
        public Connection[] newArray(int size) {
            return new Connection[size];
        }
    };
    private final Section[] sections;
    private DepartureCheckpoint departure;
    private ArrivalCheckpoint arrival;

    public Connection(Section[] sections, DepartureCheckpoint departure, ArrivalCheckpoint arrival) {
        this.sections = requireNonNull(sections, "sections");
        this.departure = requireNonNull(departure, "departure");
        this.arrival = requireNonNull(arrival, "arrival");
    }

    protected Connection(Parcel in) {
        sections = in.createTypedArray(Section.CREATOR);
        departure = ParcelUtils.readParcelable(in, DepartureCheckpoint.CREATOR);
        arrival = ParcelUtils.readParcelable(in, ArrivalCheckpoint.CREATOR);
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeTypedArray(sections, flags);
        ParcelUtils.writeParcelable(dest, departure, flags);
        ParcelUtils.writeParcelable(dest, arrival, flags);
    }

    @Override
    public int describeContents() {
        return 0;
    }

    public Section[] getSections() {
        return sections;
    }

    public Date getDepartureDate() {
        return departure.getDepartureTime();
    }

    public Date getArrivalDate() {
        return arrival.getArrivalTime();
    }
}

File: lib/common/src/main/java/ch/unstable/ost/api/model/Connection.kt
package ch.unstable.ost.api.model

import android.os.Parcelable
import java.util.*

/**
 * Create a new connection
 *
 * @param sections an array containing the sections
 */
interface Connection : Parcelable {

    val sections: List<Section>

    /**
     * Get the departure time
     *
     * @return the time of the departure
     */
    val departureDate: Date
        get() = departure.time

    /**
     * Get the arrival time
     *
     * @return the arrival time
     */
    val arrivalDate: Date
        get() = arrival.time

    /**
     * Get the departure checkpoint
     *
     *
     * The departure checkpoint of a connection is the departure checkpoint of the first section.
     *
     * @return the departure checkpoint
     */
    val departure
        get() = sections.first().departure

    /**
     * Get the arrival checkpoint
     *
     *
     * The arrival checkpoint of a connection is the arrival checkpoint of the last section.
     *
     * @return the arrival checkpoint
     */
    val arrival
        get() = sections.last().arrival

    val departureName: String
        get() = departure.station.name

    val arrivalName: String
        get() = arrival.station.name

}